server/src/main/java/org/apache/iotdb/db/mpp/common/schematree/SchemaTreeVisitor.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.common.schematree;
+
+import org.apache.iotdb.db.metadata.path.MeasurementPath;
+import org.apache.iotdb.db.metadata.path.PartialPath;
+
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.regex.Pattern;
+
+import static org.apache.iotdb.commons.conf.IoTDBConstant.MULTI_LEVEL_PATH_WILDCARD;
+import static org.apache.iotdb.commons.conf.IoTDBConstant.ONE_LEVEL_PATH_WILDCARD;
+
+public class SchemaTreeVisitor implements Iterator<MeasurementPath> {
+
+  private final SchemaNode root;
+  private final String[] nodes;
+  private final boolean isPrefixMatch;
+
+  private final int limit;
+  private final int offset;
+  private final boolean hasLimit;
+
+  private int count = 0;
+  private int curOffset = -1;
+
+  private final Deque<Integer> indexStack = new LinkedList<>();
+  private final Deque<Iterator<SchemaNode>> stack = new LinkedList<>();
+  private final Deque<SchemaNode> context = new LinkedList<>();
+
+  private SchemaMeasurementNode nextMatchedNode;
+
+  public SchemaTreeVisitor(
+      SchemaNode root, PartialPath pathPattern, int slimit, int soffset, boolean isPrefixMatch) {
+    this.root = root;
+    nodes = pathPattern.getNodes();
+    this.isPrefixMatch = isPrefixMatch;
+
+    limit = slimit;
+    offset = soffset;
+    hasLimit = slimit != 0;
+
+    indexStack.push(0);
+    stack.push(Collections.singletonList(root).iterator());
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (nextMatchedNode == null) {
+      getNext();
+    }
+    return nextMatchedNode != null;
+  }
+
+  @Override
+  public MeasurementPath next() {
+    if (!hasNext()) {
+      throw new NoSuchElementException();
+    }
+    MeasurementPath result = generateMeasurementPath();
+    nextMatchedNode = null;
+    return result;
+  }
+
+  public List<MeasurementPath> getAllResult() {
+    List<MeasurementPath> result = new LinkedList<>();
```

Comment:
Why we use a LinkedList rather than ArrayList here?

Reply:
Fixed. I thought that since there won't be random access on this list, it will be better to use LinkedList, which will avoid Array copy during add.